include/aero/array/array.h

```diff
@@ -1,7 +1,58 @@
 #ifndef AERO_ARRAY_H
 #define AERO_ARRAY_H
 
-/// This opaque type represents an array implemented in C.
+#include <aero/aero.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/// This opaque type represents an implementation of an Array
+typedef struct aero_array_data_t aero_array_data_t;
+
+/// Behaviors and data associated with an Array. All of these functions
+/// are passed the data pointer created by an Array constructor
 typedef struct aero_array_t aero_array_t;
+typedef struct aero_array_t {
+  /// pointer to the array data
+  aero_array_data_t *data_;
+  /// returns a clone (deep-copy) of an array
+  aero_array_t* (*clone)(aero_array_t *from);
+  /// frees resources associated with an array
+  void (*free)(aero_array_t *array);
+  /// copies data into an array
+  void (*copy_in)(aero_array_t *array, const aero_real_t *from);
+  /// copies data out of an array
+  void (*copy_out)(aero_array_t *array, aero_real_t *to);
+  /// returns the number of elements in an array
+  unsigned int (*size)(aero_array_t *array);
+} aero_array_behaviors_t;
+
+/// Creates an array from a number of elements and an initial value
+aero_array_t* aero_array_from_dimensions(const int number_of_elements,
```

Comment:
We could call this `aero_array_new` to give it the feel of a constructor. I understand if you prefer the "symmetry" of this naming scheme though. Also, I am predisposed to shorter names as long as they are fairly clear and unambiguous (e.g. `size` instead of `number_of_elements`). But naming is one of the most contentious topics in our field. :-)

Reply:
I agree with short names, but I thought this one could be ambiguous because size could be `number_of_elements` or `number_of_elements * sizeof(Real)`. But, if you think it would be clear what `size` means, I'm ok with this.